### PR TITLE
[Bug] Fix category error on authentication

### PR DESCRIPTION
**Pull request by oreoluwade**
#### What does this PR do?
- Fixes the disappearance of the category items on logout.
- Delegates cart queries responsibilities to the new `CartContext`
- General refactor

#### How should this be manually tested?
- Authenticate and try to add and remove items from the cart.
- Login and logout and observe that the categories no longer disappear

#### Any background context you want to provide?
- The clearing of the store on logout resulted in the category items disappearance. Also, was noticed that multiple users on the same DB could not properly query for item quantity, and so a task was carried out on the backend to fix that. The other changes on the frontend were done to fix recurring irregularities with observed user actions.

#### What are the relevant pivotal tracker stories?
- [#167557383](https://www.pivotaltracker.com/story/show/167557383)

### CODE REVIEW

File: src/components/Authentication/Login/login-form.jsx [hunk 1]
```diff
@@ -121,7 +115,6 @@ const LoginForm = ({
         const { key, user } = userLogin.token;
         addToStorage({ user: encrypt(user.id), token: key });
         handleAuthStatusChange(true);
-        client.resetStore();
```
Comment (omorilewa) [inline src/components/Authentication/Login/login-form.jsx]: @oreoluwade Do you know why this was there before? Asking so that we are sure removing it does not cause any unexpected behaviour.

File: src/components/Authentication/SignUp/signup-form.jsx [hunk 2]
```diff
@@ -96,29 +97,12 @@ const SignupForm = ({
         }
       }}
     >
-      {(createUser, { loading }) => (
+      {(mutate, { loading }) => (
```
Comment (omorilewa) [inline src/components/Authentication/SignUp/signup-form.jsx]: Is mutate more descriptive? Why the change?

File: src/components/Product/product-details.jsx [hunk 3]
```diff
@@ -94,23 +94,21 @@ const ProductDetails = ({ location, match }) => {
...
-  const { user, cartItems } = useContext(UserContext);
+  const { user } = useContext(UserContext);
+  const { cartItems } = useContext(CartContext);
+
   const [value, setValue] = useState(0);
   const [quantityToBeAdded, setQuantityToBeAdded] = useState(1);
 
-  const handleQuantityChange = event =>
-    setQuantityToBeAdded(event.target.value);
+  const handleQuantityChange = ({ target }) =>
```
Comment (omorilewa) [inline src/components/Product/product-details.jsx]: 👍

File: src/components/Authentication/Login/login-form.jsx (hunk 1 above)
Comment (oreoluwade) [reply src/components/Authentication/Login/login-form.jsx]: Yes. I put it there in the first place cos I was trying to clear all data from the store. Unfortunately, the categories are part of what is in store, so can't afford to clear the store.

File: src/components/Authentication/SignUp/signup-form.jsx (hunk 2 above)
Comment (oreoluwade) [reply src/components/Authentication/SignUp/signup-form.jsx]: Hmmm. Maybe not. Guess I have been so used to using the keyword `mutate` that I just changed that. I'd revert this.

File: src/components/Authentication/Login/login-form.jsx (hunk 1 above)
Comment (omorilewa) [reply src/components/Authentication/Login/login-form.jsx]: Oh okay, Makes sense to me,

File: src/contexts/cart-provider.jsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  useEffect(() => {
+    if (decryptedId && data.userCartItems) {
+      setState({ cartItems: data.userCartItems });
+      updateNumberOfCartItems(data.userCartItems);
+    }
+  }, [data]);
+
+  const updateNumberOfCartItems = cartItemsArray => {
+    setState({
+      numberOfCartItems: cartItemsArray.reduce((acc, next) => {
```
Comment (omorilewa) [inline src/contexts/cart-provider.jsx]: Sleek.

**Comment by omorilewa** [comment]
I am not sure what I may be doing wrong but when I 
- Signup with a new user
- Add an item to the cart

The cart button on the navbar does not update immediately, is that expected?

Additionaly, when I tried to add the lip gloss product to the cart, I get this error in the console

Screenshot 2019-07-28 at 11 12 43 PM

The first item gets added, I see this after a refresh, the second which is the lip gloss throws tha error above.

**Comment by oreoluwade** [comment]
> I am not sure what I may be doing wrong but when I
> 
> * Signup with a new user
> * Add an item to the cart
> 
> The cart button on the navbar does not update immediately, is that expected?
> 
> Additionaly, when I tried to add the lip gloss product to the cart, I get this error in the console
> 
> Screenshot 2019-07-28 at 11 12 43 PM
> 
> The first item gets added, I see this after a refresh, the second which is the lip gloss throws the error above.

There is a PR on the backend that addresses this. Please check this branch `bug/fix-multiple-results-for-product-quantity` and the PR, you'd understand.

**Comment by oreoluwade** [comment]
So temporarily, please run the api on that branch while you redo your test on the frontend. @omorilewa

**Comment by omorilewa** [comment]
Checking out to that branch fixed the isssue. I am good with approving this PR. I suggest we merge the backend one in before this goes.

Asides that solid work!

**Comment by oreoluwade** [comment]
Thank you, thank you. Yeah, the backend needs to be merged first. Also, maybe tomorrow, we'd discuss about an error for a newly signed up user. I do not think we should be returning an error for that case, but again, will hold off on the two PRs till tomorrow morning @omorilewa

**Comment by omorilewa** [comment]
Sounds good.
